homeassistant/components/text/__init__.py

```diff
@@ -0,0 +1,215 @@
+from homeassistant.const import MAX_LENGTH_STATE_STATE
+from homeassistant.core import HomeAssistant, ServiceCall
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.config_validation import (  # noqa: F401
+    PLATFORM_SCHEMA,
+    PLATFORM_SCHEMA_BASE,
+)
+from homeassistant.helpers.entity import Entity, EntityDescription
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.typing import ConfigType
+
+from .const import (
+    ATTR_MAX,
+    ATTR_MIN,
+    ATTR_MODE,
+    ATTR_PATTERN,
+    ATTR_VALUE,
+    DOMAIN,
+    SERVICE_SET_VALUE,
+)
+
+SCAN_INTERVAL = timedelta(seconds=30)
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+_LOGGER = logging.getLogger(__name__)
+
+__all__ = ["DOMAIN", "TextEntity", "TextEntityDescription", "TextMode"]
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up Text entities."""
+    component = hass.data[DOMAIN] = EntityComponent[TextEntity](
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL
+    )
+    await component.async_setup(config)
+
+    component.async_register_entity_service(
+        SERVICE_SET_VALUE,
+        {vol.Required(ATTR_VALUE): cv.string},
+        _async_set_value,
+    )
+
+    return True
+
+
+async def _async_set_value(entity: TextEntity, service_call: ServiceCall) -> None:
+    """Service call wrapper to set a new value."""
+    value = service_call.data[ATTR_VALUE]
+    if len(value) < entity.min:
+        raise HomeAssistantError(
+            f"Value {value} for {entity.name} is too short (minimum length {entity.min})"
+        )
+    if len(value) > entity.max:
+        raise HomeAssistantError(
+            f"Value {value} for {entity.name} is too long (maximum length {entity.max})"
+        )
+    if entity.pattern is not None and re.compile(entity.pattern).match(value) is None:
+        raise HomeAssistantError(
+            f"Value {value} for {entity.name} doesn't match pattern {entity.pattern}"
+        )
+    await entity.async_set_value(value)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[TextEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class TextMode(StrEnum):
+    """Modes for text entities."""
+
+    PASSWORD = "password"
+    TEXT = "text"
+
+
+@dataclass
+class TextEntityDescription(EntityDescription):
+    """A class that describes text entities."""
+
+    native_min: int | None = None
+    native_max: int | None = None
+    mode: TextMode | None = None
```

Comment:
If we set the default here, we can remove the need for allowing `None`.

```suggestion
    mode: TextMode = TextMode.TEXT
```